#10314 Better worker-side error handling in disttrial

exarkun opened the pull request:
## Scope and purpose

Fixes #10314 

This adds error handling (any, at all!) to the "workerreporter" that manages communication of test results from the disttrial worker processes back to the parent process so that such errors won't cause the whole trial process to exit without a result report.

It also changes the network protocol between workers and parent to stream error text so that text longer than the AMP limit of 64k makes it back to the parent without error.

exarkun committed:
Add a failing test for behavior of an error sending success result

exarkun committed:
Add some robustness against protocol changes

exarkun committed:
Catch errors reporting success and make them into test errors

exarkun committed:
Add a test for `addError` errors and handle them.

exarkun committed:
Add test for errors in the other TestResult methods and fix them.

exarkun committed:
Stream errors and failures from the disttrial worker back to the parent

This avoids a protocol error when the string representations exceed the length
limits imposed by the AMP protocol disttrial uses.

exarkun committed:
news fragment

exarkun committed:
Compat with older Python

exarkun committed:
Make the string too long instead of making a TypeError

exarkun committed:
Stream the expected failure error, too.

exarkun committed:
formatting

exarkun commented:
please review

exarkun committed:
Cover the last couple error cases

adiroiban commented on src/twisted/trial/_dist/stream.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        self._streams[streamId].append(chunk)
+
+    def close(self, streamId: int) -> List[str]:
```
I am a bit confused by the `.close()` as I am not normally expecting it to return the content.

But the method is well documented...so not a blocker.

Maybe consider using:

* open / start
* write / append
* finalize()

I think that finalizing is also used by crypto hash/ciphers to end the process and return something.

exarkun replied:
Maybe `finish` or `popContent`?

adiroiban replied:
I think `finish` is ok, as it should imply that no future actions are valid for the streamID.

`popContent` might lead you to think that you "flush" the current content, but you can continue to write future content.

exarkun replied:
Addressed this in the version on at https://github.com/twisted/twisted/pull/11662

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -137,15 +185,17 @@ def _buildFailure(
+        framesStreamId: int,
     ) -> Dict[str, bool]:
         """
         Add an error to the reporter.
```
Maybe extend the docstring to inform that before adding the error,
the error "content" should be previously sent via the streaming AMP interface.


Same for addFailure

adiroiban reviewed (commented):
I did a very quick review.
I am leaving the command as I don't know when I will have time to do a full review.

If no other developer will have time to review this in the next few days, I think that this should be merge.

I don't think is ok to block a merge for a PR that has tests for the new code, is well documented and the tests are green.

exarkun committed:
doc improvements related to the dist worker implementation

exarkun committed:
update to the new method name

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 10314.disttrial-error-handling

adiroiban commented on src/twisted/trial/_dist/test/test_workertrial.py:
```diff
@@ -101,7 +101,9 @@ def test_forwardCommand(self):
         main(self.fdopen)
-        self.assertIn(b"No module named 'doesntexist'", self.writeStream.getvalue())
+        # Just brazenly encode irrelevant implementation details here, why
+        # not.
```
:)

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -68,20 +79,44 @@ class WorkerProtocol(AMP):
+                        id=testCase,
+                    )
+                    try:
+                        await self._result._addErrorFallible(testCase, result)
```
Maybe have it a public interface

```suggestion
                        await self._result.addErrorFallible(testCase, result)
```

The idea is that maybe in the future we would want to run some code scanners to detect for private access.

adiroiban commented on src/twisted/trial/_dist/workerreporter.py:
```diff
@@ -56,48 +215,85 @@ def _getFrames(self, failure: Failure) -> List[str]:
+        )
 
-    def addError(self, test, error):
+    async def _addErrorFallible(self, testName: str, errorObj: TrialFailure) -> None:
```
Maybe make it public.

The docs are good enough and I hope people will not abuse it.

```suggestion
    async def addErrorFallible(self, testName: str, errorObj: TrialFailure) -> None:
```

adiroiban reviewed (commented):
It looks good. Only minor comments.

-------

exarkun committed:
Make `addErrorFallible` public

It's still on a class in a private package so this doesn't make it public to
users outside of Twisted (or even outside of disttrial).  It does make it
public to `WorkerProtocol` which needs to use it.

adiroiban commented on src/twisted/trial/test/erroneous.py:
```diff
@@ -109,12 +124,26 @@ def test_fail(self) -> defer.Deferred[None]:
         """
         raise Exception(self.text)
 
+    def test_exceptionGreaterThan64k(self) -> None:
```
Maybe add an info that this test is actually executed as part `[test_workerreporter.py](https://github.com/twisted/twisted/pull/11644/files#diff-20c8b908f85e8394203296722e0e1dc7013d9c1fdf8d4d86e7e354634da7f3e0)

the top docstring of this module talks about `L{twisted.trial.test.test_tests}`

My preference is to reduce the size of these external files... and for example have one separate file for each test

Or better, each actual tests will generate a package inside temp and will run the tests from there.

----- 

but I ok not to have the test code auto-generated as part of this PR.

but better docstring would help.

exarkun replied:
Some of the other tests in this module are also now used by `twisted.trial._dist.test` so I added a note about this to the module docstring.

I think that whatever the eventual factoring, there is still going to have to be sharing between `twisted.trial.test` and `twisted.trial._dist.test` because we want to be sure that trial and disttrial have analogous behaviors for some set of inputs (test cases).  They should share the definition of those inputs so that we know they're really the same.

adiroiban commented on src/twisted/trial/test/erroneous.py:
```diff
@@ -109,12 +124,26 @@ def test_fail(self) -> defer.Deferred[None]:
 
-    def test_exception(self):
+    def test_failGreaterThan64k(self) -> defer.Deferred[None]:
+        """
+        A test which fails in the callback of the returned L{defer.Deferred}
+        with a very long string.
+        """
```
```suggestion
        """
        A test which returns a L{Failure} in the callback of the returned L{defer.Deferred}
        with a very long string.
        
        Note: there is a separate test for checking the case in which an
        exception is raised.
        
        This is triggered via L{test_addFailureGreaterThan64k}.
        """
```

exarkun replied:
Hmm.  I'm not sure about this.  The last callback on the returned `Deferred` doesn't do `return Failure(...)` but `self.fail(...)`.  It will actually raise an exception.  Of course, this is turned into a `Failure` by `Deferred` so I'm not sure it's possible to tell the difference between these cases (except the raised exception might have more stack frames in its traceback).

adiroiban reviewed (commented):
I left a few other minor comments.

I still have to review - `_runErrorTest` and `WorkerProtocol.run`

I will try to continue the review later today.

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -68,20 +79,44 @@ class WorkerProtocol(AMP):
+            suite = TrialSuite([case], self._forceGarbageCollection)
+            suite.run(self._result)
+
+            allSucceeded = True
```
do we need this to be inside the `gatherReportingResults()` context manager?

exarkun replied:
nope, not at all, and it's better outside, thanks.

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -68,20 +79,44 @@ class WorkerProtocol(AMP):
+            suite.run(self._result)
+
+            allSucceeded = True
+            for (success, result) in await DeferredList(results, consumeErrors=True):
+                if not success:
+                    allSucceeded = False
```
Maybe reduce the indentation and add explicit handling of the success case.

```suggestion
            for (success, result) in await DeferredList(results, consumeErrors=True):
                if success:
                    # Test execution was successful and there is no need for extra
                    # reporting.
                    continue

                # Report the failed test.
                allSucceeded = False
```

adiroiban commented on src/twisted/trial/_dist/worker.py:
```diff
@@ -68,20 +79,44 @@ class WorkerProtocol(AMP):
+                    )
+                    try:
+                        await self._result.addErrorFallible(testCase, result)
+                    except BaseException:
```
```suggestion
                    except BaseException:
                        # We failed to report the failure to the centralized process.
                        # Just add extra local logging, as there is not much else we can do.
```

adiroiban reviewed (approved):
Looks good.

Only a medium comment regarding the usage of context manager.

But I think it can be fixed and merged without another round of reviews.

exarkun committed:
note an additional user of this module

exarkun committed:
move reporting results processing out of the context manager

All reporting results must have happened by the time `suite.run` returns so
there's no reason to process them inside the context manager.  And processing
them in the context manager means that if any code somehow starts a _new_
reporting operation the list we passed to DeferredList will be mutated (and
presumably then ignore).  Ending the context manager turns that case into a
hard error, instead.

exarkun committed:
Flip the success/failure cases for less indentation

exarkun committed:
add a comment explaining this error handling/logging case

exarkun committed:
Merge remote-tracking branch 'origin/trunk' into 10314.disttrial-error-handling
